src/machine/machine_k210.go

```diff
@@ -0,0 +1,642 @@
+	}
+
+	if f >= FUNC_GPIO0 && f <= FUNC_GPIO7 {
+		gpioPin := uint8(f - FUNC_GPIO0)
+
+		if high {
+			kendryte.GPIO.DATA_OUTPUT.SetBits(1 << gpioPin)
+		} else {
+			kendryte.GPIO.DATA_OUTPUT.ClearBits(1 << gpioPin)
+		}
+	} else if f >= FUNC_GPIOHS0 && f <= FUNC_GPIOHS31 {
+		gpioPin := uint8(f - FUNC_GPIOHS0)
+
+		if high {
+			kendryte.GPIOHS.OUTPUT_VAL.SetBits(1 << gpioPin)
+		} else {
+			kendryte.GPIOHS.OUTPUT_VAL.ClearBits(1 << gpioPin)
+		}
+	}
+}
+
+// Get returns the current value of a GPIO pin.
+func (p Pin) Get() bool {
+
+	// Check if the current pin's FPIOA function is either GPIO or GPIOHS.
+	f := p.FPIOAFunction()
+	if f < FUNC_GPIOHS0 || f > FUNC_GPIO7 {
+		return false // The pin is not configured as GPIO or GPIOHS.
+	}
+
+	var val uint32
+	if f >= FUNC_GPIO0 && f <= FUNC_GPIO7 {
+		gpioPin := uint8(f - FUNC_GPIO0)
+		val = kendryte.GPIO.DATA_INPUT.Get() & (1 << gpioPin)
+	} else if f >= FUNC_GPIOHS0 && f <= FUNC_GPIOHS31 {
+		gpioPin := uint8(f - FUNC_GPIOHS0)
+		val = kendryte.GPIOHS.INPUT_VAL.Get() & (1 << gpioPin)
+	}
+	return (val > 0)
+}
+
+// Callbacks to be called for GPIOHS pins configured with SetInterrupt.
+var pinCallbacks [32]func(Pin)
+
+// SetInterrupt sets an interrupt to be executed when a particular pin changes
+// state.
+//
+// You can pass a nil func to unset the pin change interrupt. If you do so,
+// the change parameter is ignored and can be set to any value (such as 0).
+// If the pin is already configured with a callback, you must first unset
+// this pins interrupt before you can set a new callback.
+func (p Pin) SetInterrupt(change PinChange, callback func(Pin)) error {
+
+	// Check if the pin is a GPIOHS pin.
+	f := p.FPIOAFunction()
+	if f < FUNC_GPIOHS0 || f > FUNC_GPIOHS31 {
+		return ErrInvalidDataPin
+	}
+
+	gpioPin := uint8(f - FUNC_GPIOHS0)
+
+	// Clear all interrupts.
+	kendryte.GPIOHS.RISE_IE.ClearBits(1 << gpioPin)
+	kendryte.GPIOHS.FALL_IE.ClearBits(1 << gpioPin)
+	kendryte.GPIOHS.HIGH_IE.ClearBits(1 << gpioPin)
+	kendryte.GPIOHS.LOW_IE.ClearBits(1 << gpioPin)
+
+	// Clear all the pending bits for this pin.
+	kendryte.GPIOHS.RISE_IP.SetBits(1 << gpioPin)
+	kendryte.GPIOHS.FALL_IP.SetBits(1 << gpioPin)
+	kendryte.GPIOHS.HIGH_IP.SetBits(1 << gpioPin)
+	kendryte.GPIOHS.LOW_IP.SetBits(1 << gpioPin)
+
+	if callback == nil {
+		if pinCallbacks[gpioPin] != nil {
+			pinCallbacks[gpioPin] = nil
+		}
+		return nil
+	}
+
+	if pinCallbacks[gpioPin] != nil {
+		// The pin was already configured.
+		// To properly re-configure a pin, unset it first and set a new
+		// configuration.
+		return ErrNoPinChangeChannel
+	}
+
+	pinCallbacks[gpioPin] = callback
+
+	// Enable interrupts.
+	if change&PinRising != 0 {
+		kendryte.GPIOHS.RISE_IE.SetBits(1 << gpioPin)
+	}
+	if change&PinFalling != 0 {
+		kendryte.GPIOHS.FALL_IE.SetBits(1 << gpioPin)
+	}
+	if change&PinHigh != 0 {
+		kendryte.GPIOHS.HIGH_IE.SetBits(1 << gpioPin)
+	}
+	if change&PinLow != 0 {
+		kendryte.GPIOHS.LOW_IE.SetBits(1 << gpioPin)
+	}
```

Comment:
What are the semantics for `PinHigh` and `PinLow`? The interrupt keeps being called continuously as long as the line is high (or low)?
I intentionally did not implement this on other chips as it seemed less useful to me and I wasn't sure how it should be defined exactly in a portable way.

Reply:
> The interrupt keeps being called continuously as long as the line is high (or low)?

Yes, I agree it is not very useful, I will remove it.